schema/pgsql/schema.sql

```diff
@@ -0,0 +1,1904 @@
+-- Icinga DB | (c) 2021 Icinga GmbH | GPLv2+
+
+CREATE DOMAIN bytea20 AS bytea CONSTRAINT exactly_20_bytes_long CHECK ( VALUE IS NULL OR octet_length(VALUE) = 20 );
+CREATE DOMAIN bytea16 AS bytea CONSTRAINT exactly_16_bytes_long CHECK ( VALUE IS NULL OR octet_length(VALUE) = 16 );
+CREATE DOMAIN bytea4 AS bytea CONSTRAINT exactly_4_bytes_long CHECK ( VALUE IS NULL OR octet_length(VALUE) = 4 );
+
+CREATE DOMAIN biguint AS bigint CONSTRAINT positive CHECK ( VALUE IS NULL OR 0 <= VALUE );
+CREATE DOMAIN uint AS bigint CONSTRAINT between_0_and_4294967295 CHECK ( VALUE IS NULL OR VALUE BETWEEN 0 AND 4294967295 );
+CREATE DOMAIN smalluint AS int CONSTRAINT between_0_and_65535 CHECK ( VALUE IS NULL OR VALUE BETWEEN 0 AND 65535 );
+CREATE DOMAIN tinyuint AS smallint CONSTRAINT between_0_and_255 CHECK ( VALUE IS NULL OR VALUE BETWEEN 0 AND 255 );
```

Comment:
Why should we do that? Because of storage? I don't think you want to keep track of possible separate values for these columns for MySQL and PostgreSQL. It always has to be the same. So yes, we need those types.

Reply:
One would have to look into the individual columns: what is stored in there and what's the best type supported by the database to store this. IMHO this might very well result in different types for MySQL and PostgreSQL.

Reply:
I do not understand what you are trying to say. The types are there so that there are no possible different values for these types in MySQL and PostgreSQL.

Reply:
I'm trying to say that columns in MySQL are probably `smallint unsigned` because that was the best-fitting type that MySQL had to offer, not because the values we want to store use the full `uint16` range. So the best type PostgreSQL has to offer for the same column might as well be a `smallint` or maybe also a `bit(n)` (maybe that would make sense if we store a bitset) or something totally different.

The domains introduced here look to me like trying to make it look like MySQL at all cost.

Reply:
> The domains introduced here look to me like trying to make it look like MySQL at all cost.

Of course, that's the whole point. But it's about the behavior! If a value is wrong it's equally wrong in both databases. But without something like that there might be totally different values accepted. And since MySQL doesn't support custom data types, it's just the way it is.

Reply:
I'm sorry, but then I don't get the point. Let's take the very first column using one of these domains as an example: `host.max_check_attempts`.  Does it make a meaningful difference if this is stored as `int` or as `bigint` constrained to unsigned 32 bit values? Both support all sane values for that config option and neither allows the full range Icinga 2 would accept for this value. Therefore, I'd go with the simpler variant and just use `int` (or possibly even an unconstrained `bigint` to allow everything allowed by Icinga 2, even if that configuration wouldn't make any sense).

Reply:
What matters here is not what makes sense for the possible configuration values, but what is set in the MySQL schema. The maximum value there is **4294967295**. This has to be the same in PostgreSQL.

Reply:
Why is the allowed range now defined by the MySQL schema and not by Icinga 2? After all, the row represents an Icinga 2 object.

Reply:
I can only repeat myself as this has nothing to do with this discussion. Same schemes are a must. No matter what permissible values we have in MySQL. These must be exactly the same in PostgreSQL.